runtime/src/main/java/org/corfudb/security/tls/ReloadableKeyManager.java

```diff
@@ -0,0 +1,123 @@
+package org.corfudb.security.tls;
+
+import lombok.extern.slf4j.Slf4j;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.X509KeyManager;
+import java.net.Socket;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.X509Certificate;
+
+@Slf4j
+public class ReloadableKeyManager implements X509KeyManager {
+
+    private final TlsUtils.CertStoreConfig.KeyStoreConfig keyStoreConfig;
+
+    private X509KeyManager keyManager;
+
+    private long keysLastModifiedTime;
+
+    public ReloadableKeyManager(TlsUtils.CertStoreConfig.KeyStoreConfig keyStoreConfig) {
+        this.keyStoreConfig = keyStoreConfig;
+        this.keysLastModifiedTime = 0;
+        reloadKeyStoreWrapper();
+    }
+
+    @Override
+    public String[] getClientAliases(String keyType, Principal[] issuers) {
+        reloadKeyStoreWrapper();
+        return keyManager.getClientAliases(keyType, issuers);
+    }
+
+    @Override
+    public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
+        reloadKeyStoreWrapper();
+        return keyManager.chooseClientAlias(keyType, issuers, socket);
+    }
+
+    @Override
+    public String[] getServerAliases(String keyType, Principal[] issuers) {
+        reloadKeyStoreWrapper();
+        return keyManager.getServerAliases(keyType, issuers);
+    }
+
+    @Override
+    public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+        reloadKeyStoreWrapper();
+        return keyManager.chooseServerAlias(keyType, issuers, socket);
+    }
+
+    @Override
+    public X509Certificate[] getCertificateChain(String alias) {
+        reloadKeyStoreWrapper();
+        return keyManager.getCertificateChain(alias);
+    }
+
+    @Override
+    public PrivateKey getPrivateKey(String alias) {
+        reloadKeyStoreWrapper();
+        return keyManager.getPrivateKey(alias);
+    }
+
+    private boolean hasKeysChanged() {
+        long lastModified = keyStoreConfig.getKeyStoreFile().toFile().lastModified();
+        if (lastModified != keysLastModifiedTime) {
+            keysLastModifiedTime = lastModified;
+            return true;
+        }
+        return false;
+    }
+
+    private void reloadKeyStoreWrapper() {
+        try {
+            reloadKeyStore();
+        } catch (SSLException e) {
+            String message = "Unable to reload key store " + keyStoreConfig.getKeyStoreFile() + ".";
+            throw new RuntimeException(message, e);
```

Comment:
Please throw IllegalStateException. Throwing raw runtime exceptions is not reccomended

Reply:
Done.